### PR TITLE
Solidity friendly precise proofs

### PR DESCRIPTION
**Pull request by kosecki123**
This PR enables precise proofs generation which is possible to proof using an on-chain verifier.

Changes:
- hashing/packing function changed from `sha3` with string concatenation to `web3.utils.soliditySha3` 
- string values are now base64 encoded before hashing - this allows to support number-as-string inputs for e.g `"1.5"` which by default are decoded to number/BN and causing exceptions
- hashes have `0x` prefixes
- more performant scheme sort and find

**BREAKING: This PR breaks the compatibility with previously created proofs.**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,230 +1,252 @@
...
+
+  export const hash = (...params: any[]): string => {
+    return utils.soliditySha3(...params);
+  };
+
+  const getSalt = (length: number, encoding: string = "base64") => {
+    return crypto
+      .randomBytes(length)
+      .toString(encoding)
+      .slice(0, length);
```
Comment (ngyam) [inline src/index.ts]: Let's cut the .slice(0, length) from the end for more entropy.

Comment (ngyam) [reply src/index.ts]: and rename length to bytelength then

**Comment by ngyam** [comment]
ping @kosecki123

**Comment by kosecki123** [comment]
@ngyam resolved your proposal on using byteLength rather than string length

**Pull request merged**